File: src/main/java/com/chicman/view/ProductItem.java
package com.chicman.view;

import com.chicman.ChicManUI;
import com.chicman.model.Product;
import com.chicman.utility.Messages;
import com.chicman.utility.Utils;
import com.vaadin.icons.VaadinIcons;
import com.vaadin.server.StreamResource;
import com.vaadin.ui.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ProductItem extends CssLayout {
    private final Logger log = LoggerFactory.getLogger(ProductItem.class);

    private final ChicManUI eventHandler;
    private final Product product;
    private final Image productImg;
    private final Button likeBtn;
    private StreamResource imgResource;

    private boolean showFlag;
    private boolean filteredByDialFlag;
    private boolean filteredByBrandFlag;
    private boolean filteredByDiscountFlag;
    private boolean filteredByStrapFlag;
    private boolean filteredByDisplayFlag;
    private boolean filteredByDialColorFlag;
    private boolean filteredByStrapColorFlag;

    public ProductItem(ChicManUI eventHandler, Product product) {
        this.eventHandler = eventHandler;
        this.product = product;
        this.productImg = new Image();
        this.likeBtn = new Button();

        init();
        setEvent();
        setFavourite2ProductItem();
    }

    private void setFavourite2ProductItem() {
        likeBtn.removeStyleName("like-btn-active");
        for (Product favouritItem: eventHandler.getSessionFavourite().getFavouriteList()) {
            if (favouritItem.getProductId().intValue() == product.getProductId().intValue()) {
                likeBtn.addStyleName("like-btn-active");
            }
        }
    }

    private void init() {
        setStyleName("product_card_item");

        StreamResource.StreamSource imageSource = new Utils.ImageSource(product.getImage());
        imgResource = new StreamResource(imageSource, "product_" + product.getProductId() + ".png");
        productImg.setSource(imgResource);
        productImg.setWidth("300px");
        addComponent(productImg);

        VerticalLayout verticalLayout = new VerticalLayout();
        verticalLayout.setWidthUndefined();
        verticalLayout.setMargin(false);
        addComponent(verticalLayout);

        Label productDetails = new Label();
        productDetails.setValue(product.getProductDetails());
        productDetails.addStyleName("item_details");
        productDetails.setWidth("300px");
        verticalLayout.addComponent(productDetails);

        CssLayout priceContainer = new CssLayout();
        verticalLayout.addComponent(priceContainer);

        Label productPrice = new Label();
        productPrice.addStyleName("card_item_price");
        productPrice.setValue(Messages.get("price.currency") + Utils.float2Text(product.getProductPrice()));
        if (product.getProductDiscount() > 0) {
            priceContainer.addComponent(productPrice);
        }

        Label productPriceSale = new Label();
        productPriceSale.addStyleName("card_item_price_sale");
        productPriceSale.setValue(Messages.get("price.currency") + Utils.float2Text(product.getProductPriceSale()));
        priceContainer.addComponent(productPriceSale);

        Label productDiscount = new Label();
        productDiscount.addStyleName("card_item_discount");
        productDiscount.setValue(Utils.float2Text(product.getProductDiscount()) + "% off");
        if (product.getProductDiscount() > 0) {
            priceContainer.addComponent(productDiscount);
        }

        HorizontalLayout ratingContainer = new HorizontalLayout();
        ratingContainer.addStyleName("rating_container");
        ratingContainer.setWidth("100%");
        ratingContainer.setMargin(false);
        ratingContainer.setSpacing(false);
        verticalLayout.addComponent(ratingContainer);

        Label productRate = new Label();
        productRate.addStyleName("cardItemRating");
        Utils.ratingClassified(productRate, product);
        ratingContainer.addComponent(productRate);
        ratingContainer.setExpandRatio(productRate, 1);

        likeBtn.addStyleName("borderless");
        likeBtn.addStyleName("icon-only");
        likeBtn.addStyleName("active-noborder");
        likeBtn.addStyleName("like-btn");
        likeBtn.setIcon(VaadinIcons.HEART);
        ratingContainer.addComponent(likeBtn);
    }

    private void setEvent() {
        productImg.addClickListener(e -> eventHandler.exploreProduct(product, imgResource));
        likeBtn.addClickListener(e -> checkLikeBtnActive());
    }

    private void checkLikeBtnActive() {
        if (likeBtn.getStyleName().contains("like-btn-active")) {
            // Move this action to sub-menu of user menu.
//            likeBtn.removeStyleName("like-btn-active");
//            eventHandler.cancelFavouriteProduct(product);
        } else {
            likeBtn.addStyleName("like-btn-active");

            for (Product favouriteItem: eventHandler.getSessionFavourite().getFavouriteList()) {
                if (favouriteItem.getProductId().intValue() == product.getProductId().intValue()) {
                    return;
                }
            }

            eventHandler.favouriteProduct(product);
        }
    }

    public void hideItem() {
        addStyleNames("hide_card");
    }

    public void showItem() {
        removeStyleName("hide_card");
    }

    public boolean isShow() {
        return showFlag;
    }

    public void setShow(boolean showFlag) {
        this.showFlag = showFlag;
    }

    public boolean isFilteredByDial() {
        return filteredByDialFlag;
    }

    public void setFilteredByDial(boolean filteredByDialFlag) {
        this.filteredByDialFlag = filteredByDialFlag;
    }

    public boolean isFilteredByBrand() {
        return filteredByBrandFlag;
    }

    public void setFilteredByBrand(boolean filteredByBrandFlag) {
        this.filteredByBrandFlag = filteredByBrandFlag;
    }

    public boolean isFilteredByDiscount() {
        return filteredByDiscountFlag;
    }

    public void setFilteredByDiscount(boolean filteredByDiscountFlag) {
        this.filteredByDiscountFlag = filteredByDiscountFlag;
    }

    public boolean isFilteredByStrap() {
        return filteredByStrapFlag;
    }

    public void setFilteredByStrap(boolean filteredByStrapFlag) {
        this.filteredByStrapFlag = filteredByStrapFlag;
    }

    public boolean isFilteredByDisplay() {
        return filteredByDisplayFlag;
    }

    public void setFilteredByDisplay(boolean filteredByDisplayFlag) {
        this.filteredByDisplayFlag = filteredByDisplayFlag;
    }

    public boolean isFilteredByDialColor() {
        return filteredByDialColorFlag;
    }

    public void setFilteredByDialColor(boolean filteredByDialColorFlag) {
        this.filteredByDialColorFlag = filteredByDialColorFlag;
    }

    public boolean isFilteredByStrapColor() {
        return filteredByStrapColorFlag;
    }

    public void setFilteredByStrapColor(boolean filteredByStrapColorFlag) {
        this.filteredByStrapColorFlag = filteredByStrapColorFlag;
    }

    public Product getProduct() {
        return product;
    }

    public Button getLikeBtn() {
        return likeBtn;
    }
}

File: src/main/java/com/chicman/view/ProductItem.kt
package com.chicman.view

import com.chicman.ChicManUI
import com.chicman.model.Product
import com.chicman.utility.Messages
import com.chicman.utility.Utils
import com.chicman.utility.Utils.ImageSource
import com.vaadin.icons.VaadinIcons
import com.vaadin.server.StreamResource
import com.vaadin.ui.*
import org.slf4j.LoggerFactory

class ProductItem(private val eventHandler: ChicManUI, val product: Product) : CssLayout() {

    private val log = LoggerFactory.getLogger(ProductItem::class.java)

    private val productImg: Image = Image()
    private var imgResource: StreamResource? = null
    val likeBtn: Button = Button()

    var isShow = false
    var isFilteredByDial = false
    var isFilteredByBrand = false
    var isFilteredByDiscount = false
    var isFilteredByStrap = false
    var isFilteredByDisplay = false
    var isFilteredByDialColor = false
    var isFilteredByStrapColor = false

    private fun setFavourite2ProductItem() {
        likeBtn.removeStyleName("like-btn-active")
        eventHandler.sessionFavourite.favourites.forEach {
            if (it.productId.toInt() == product.productId.toInt()) {
                likeBtn.addStyleName("like-btn-active")
            }
        }
    }

    private fun init() {
        styleName = "product_card_item"
        val imageSource: StreamResource.StreamSource = ImageSource(product.image)
        imgResource = StreamResource(imageSource, "product_" + product.productId + ".png")
        productImg.source = imgResource
        productImg.setWidth("300px")
        addComponent(productImg)
        val verticalLayout = VerticalLayout()
        verticalLayout.setWidthUndefined()
        verticalLayout.setMargin(false)
        addComponent(verticalLayout)
        val productDetails = Label()
        productDetails.value = product.productDetails
        productDetails.addStyleName("item_details")
        productDetails.setWidth("300px")
        verticalLayout.addComponent(productDetails)
        val priceContainer = CssLayout()
        verticalLayout.addComponent(priceContainer)
        val productPrice = Label()
        productPrice.addStyleName("card_item_price")
        productPrice.value = Messages.get("price.currency") + Utils.float2Text(product.productPrice)
        if (product.productDiscount > 0) {
            priceContainer.addComponent(productPrice)
        }
        val productPriceSale = Label()
        productPriceSale.addStyleName("card_item_price_sale")
        productPriceSale.value = Messages.get("price.currency") + Utils.float2Text(product.productPriceSale)
        priceContainer.addComponent(productPriceSale)
        val productDiscount = Label()
        productDiscount.addStyleName("card_item_discount")
        productDiscount.value = Utils.float2Text(product.productDiscount) + "% off"
        if (product.productDiscount > 0) {
            priceContainer.addComponent(productDiscount)
        }
        val ratingContainer = HorizontalLayout()
        ratingContainer.addStyleName("rating_container")
        ratingContainer.setWidth("100%")
        ratingContainer.setMargin(false)
        ratingContainer.isSpacing = false
        verticalLayout.addComponent(ratingContainer)
        val productRate = Label()
        productRate.addStyleName("cardItemRating")
        Utils.ratingClassified(productRate, product)
        ratingContainer.addComponent(productRate)
        ratingContainer.setExpandRatio(productRate, 1f)
        likeBtn.addStyleName("borderless")
        likeBtn.addStyleName("icon-only")
        likeBtn.addStyleName("active-noborder")
        likeBtn.addStyleName("like-btn")
        likeBtn.icon = VaadinIcons.HEART
        ratingContainer.addComponent(likeBtn)
    }

    private fun setEvent() {
        productImg.addClickListener { eventHandler.exploreProduct(product, imgResource) }
        likeBtn.addClickListener { checkLikeBtnActive() }
    }

    private fun checkLikeBtnActive() {
//        if (likeBtn.styleName.contains("like-btn-active")) {
            // Move this action to sub-menu of user menu.
//            likeBtn.removeStyleName("like-btn-active");
//            eventHandler.cancelFavouriteProduct(product);
//        } else {

        if (!likeBtn.styleName.contains("like-btn-active")) {
            likeBtn.addStyleName("like-btn-active")
            for (favouriteItem in eventHandler.sessionFavourite.favourites) {
                if (favouriteItem.productId.toInt() == product.productId.toInt()) {
                    return
                }
            }
            eventHandler.favouriteProduct(product)
        }
    }

    fun hideItem() {
        addStyleNames("hide_card")
    }

    fun showItem() {
        removeStyleName("hide_card")
    }

    init {
        init()
        setEvent()
        setFavourite2ProductItem()
    }
}